gplugins/design_recipe/DesignRecipe.py

```diff
@@ -1,81 +1,151 @@
 from __future__ import annotations
 
-from typing import List, Optional
+import json
+import os
+import pickle
+from collections.abc import Callable
+from pathlib import Path, PosixPath, WindowsPath
+
+import pydantic
 from gdsfactory import Component
-from gdsfactory.pdk import LayerStack, get_layer_stack
+from gdsfactory.config import logger
 from gdsfactory.path import hashlib
+from gdsfactory.pdk import LayerStack, get_layer_stack
+from gdsfactory.typings import ComponentFactory
+from pydantic import BaseModel
 
 import gplugins.design_recipe as dr
 
 
+class Setup:
+    """
+    A dynamic class to store any setup information
+
+    This class allows designers to arbitrarily add setup information.
+    """
+
+    def __init__(self, **kwargs):
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    def __eq__(self, other):
+        if not isinstance(other, Setup):
+            # Don't attempt to compare against unrelated types
+            return NotImplemented
+        return self.__dict__ == other.__dict__
+
+
 class DesignRecipe:
     """
     A DesignRecipe represents a flow of operations on GDSFactory components,
     with zero or more dependencies. Note that dependencies are assumed to be independent,
-    dependent dependencies should be nested. When `eval()`ed, A DesignRecipe `eval()`s 
+    dependent dependencies should be nested. When `eval()`ed, A DesignRecipe `eval()`s
     its dependencies if they they've become stale,
     and optionally executes some tool-specific functionality.
     For example,an FdtdDesignRecipe might simulate its `component` in
     Lumerial FDTD to extract its s-parameters.
-    """
 
-    # This `DesignRecipe`s dependencies. These are assumed to be independent
-    dependencies: dr.ConstituentRecipes
-
-    # the hash of the system last time eval() was executed
-    last_hash: int
 
-    # The component this DesignRecipe operates on. This is not necessarily
-    # the same `component` refered to in the `dependencies` recipes.
-    component: Optional[Component] = None
+    Attributes:
+        dependencies: This `DesignRecipe`s dependencies, assumed to be independent
+        last_hash: The hash of the system last time eval() was executed.
+                   This is used to determine whether the recipe's configuration
+                   has changed, meaning it must be re-eval'ed.
+        cell: GDSFactory layout cell or component. This is not necessarily the
+              same `component` referred to in the `dependencies` recipes.
+        layer_stack: PDK layerstack
+        dirpath: Root directory where recipes runs are stored
+        recipe_dirpath: Recipe directory where recipe results are stored.
+                        This is only created upon eval of the recipe.
+        run_convergence: Run convergence if True. Accurate simulations come from
```

Comment:
"convergence [testing]"?

Reply:
Design recipes run simulations and it's best practices to run convergence testing to ensure accurate simulation results. Unfortunately, convergence configs can be quite different depending on the simulator. Generalizing a convergence config can be in a future scope.

Reply:
this was more a question about the comment- just 'convergence' as a verb is a little confusing